oauth2/flowctx/encoding.go

```diff
@@ -0,0 +1,92 @@
+// Copyright © 2023 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package flowctx
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"github.com/pkg/errors"
+
+	"github.com/ory/hydra/v2/jwk"
+)
+
+// Decode decodes the given string to a value.
+func Decode[T any](ctx context.Context, cipher *jwk.AEAD, encoded string) (*T, error) {
+	plaintext, err := cipher.Decrypt(ctx, encoded)
+	if err != nil {
+		return nil, err
+	}
+
+	rawBytes, err := gzip.NewReader(bytes.NewReader(plaintext))
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = rawBytes.Close() }()
+
+	var val T
+	if err = json.NewDecoder(rawBytes).Decode(&val); err != nil {
+		return nil, err
+	}
+
+	return &val, nil
+}
+
+// Encode encodes the given value to a string.
+func Encode(ctx context.Context, cipher *jwk.AEAD, val any) (s string, err error) {
+	// Steps:
+	// 1. Encode to JSON
+	// 2. GZIP
+	// 3. Encrypt with AEAD (AES-GCM) + Base64 URL-encode
```

Comment:
1. XChaCha-Poly1305 please
2. Base64 is quite lengthy due to the limited character set. Have we investigated other url-safe encoding mechanisms?

Reply:
I don't see a better option to Base64.